DA-1067: Add gsi tutorial for CrewAI

VirajAgarwal-couchbase opened the pull request:
- Updated GSI & FTS tutorial for CrewAI

VirajAgarwal1 committed:
update: made dirs for fts and gsi

VirajAgarwal1 committed:
update: dependencies of fts tutorial

VirajAgarwal1 committed:
update: frontmatter for fts

VirajAgarwal1 committed:
update: intro of fts tutorial

VirajAgarwal1 committed:
add: tutorial and frontmatter for gsi

VirajAgarwal1 committed:
update: execution results for fts and gsi tutorials

prajwal-pai77 commented on crewai/gsi/frontmatter.md:
```diff
@@ -0,0 +1,21 @@
+---
+# frontmatter
+path: "/tutorial-crewai-couchbase-rag-using-global-secondary-indexes"
```
For standardizing, keep it -with-global-secondary-index

prajwal-pai77 commented on crewai/gsi/.env.sample:
```diff
@@ -0,0 +1,12 @@
+CB_PASSWORD=""
+CB_BUCKET_NAME=""
+
+INDEX_NAME=""
```
I believe the index_name is not needed for gsi, as even if index name is not passed, by default langchain creates one

prajwal-pai77 commented on crewai/gsi/RAG_with_Couchbase_and_CrewAI.ipynb:
```diff
@@ -0,0 +1,2304 @@
+    "CB_USERNAME = os.getenv('CB_USERNAME') or input(\"Enter Couchbase username (default: Administrator): \") or 'Administrator'\n",
+    "CB_PASSWORD = os.getenv('CB_PASSWORD') or getpass.getpass(\"Enter Couchbase password (default: password): \") or 'password'\n",
+    "CB_BUCKET_NAME = os.getenv('CB_BUCKET_NAME') or input(\"Enter bucket name (default: vector-search-testing): \") or 'vector-search-testing'\n",
+    "INDEX_NAME = os.getenv('INDEX_NAME') or input(\"Enter index name (default: vector_search_crew): \") or 'vector_search_crew'\n",
```
Pls remove this

prajwal-pai77 commented on crewai/gsi/RAG_with_Couchbase_and_CrewAI.ipynb:
```diff
@@ -0,0 +1,2304 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this guide, we will walk you through building a powerful semantic search engine using [Couchbase](https://www.couchbase.com) as the backend database and [CrewAI](https://github.com/crewAIInc/crewAI) for agent-based RAG operations. CrewAI allows us to create specialized agents that can work together to handle different aspects of the RAG workflow, from document retrieval to response genera ...
```
Pls point link to developer portal

prajwal-pai77 commented on crewai/gsi/RAG_with_Couchbase_and_CrewAI.ipynb:
```diff
@@ -0,0 +1,2304 @@
+    }
+   ],
+   "source": [
+    "def check_query_service(cluster):\n",
```
Nit: Have we tried if this works with capella?

VirajAgarwal-couchbase replied:
I removed this part. Because it _does_ work.... But I cannot confirm if it ensures that GSI service will work.

prajwal-pai77 commented on crewai/gsi/RAG_with_Couchbase_and_CrewAI.ipynb:
```diff
@@ -0,0 +1,2304 @@
+    "   - Waits 2 seconds for collection to be ready\n",
+    "\n",
+    "Additional Tasks:\n",
+    "- Creates primary index on collection for query performance\n",
```
Pls remove this, primary index is created by langchain anyways

prajwal-pai77 commented on crewai/gsi/RAG_with_Couchbase_and_CrewAI.ipynb:
```diff
@@ -0,0 +1,2304 @@
+    "\n",
+    "        # Ensure primary index exists\n",
+    "        try:\n",
+    "            cluster.query(f\"CREATE PRIMARY INDEX IF NOT EXISTS ON `{bucket_name}`.`{scope_name}`.`{collection_name}`\").execute()\n",
```
Remove this

prajwal-pai77 commented on crewai/gsi/RAG_with_Couchbase_and_CrewAI.ipynb:
```diff
@@ -0,0 +1,2304 @@
+    "\n",
+    "This approach offers several benefits:\n",
+    "1. Memory Efficiency: Processing in smaller batches prevents memory overload\n",
+    "2. Error Handling: If an error occurs, only the current batch is affected\n",
```
Pls remove this, as this is not true

VirajAgarwal-couchbase replied:
My bad. I had some misconceptions.

prajwal-pai77 commented on crewai/gsi/RAG_with_Couchbase_and_CrewAI.ipynb:
```diff
@@ -0,0 +1,2304 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Creating a Vector Search Tool\n",
```
Nit: Maybe also have a cell showing users how to create composite index, no need to execute it

VirajAgarwal-couchbase replied:
added a comment for the same since, it just a matter of using a different `IndexType`

prajwal-pai77 commented on crewai/fts/frontmatter.md:
```diff
@@ -1,8 +1,8 @@
-path: "/tutorial-crewai-couchbase-rag"
-title: Retrieval-Augmented Generation (RAG) with Couchbase and CrewAI
-short_title: RAG with Couchbase and CrewAI
+path: "/tutorial-crewai-couchbase-rag-using-fts"
```
Same: use -with-fts

prajwal-pai77 reviewed (changes requested):
Couple of changes & should be good to go

VirajAgarwal1 committed:
update: frontmatter paths for gsi and fts

VirajAgarwal1 committed:
update: link to gsi tutorial in fts to devportal link

VirajAgarwal1 committed:
update: comments by Prajwal

- link to fts tutorial in gsi to devportal link
- removed INDEX_NAME env var
- removed Query Service check since, it isn't confirmed if it is wokring as expected
- rmeoved setup task of creating a primary index
- removed vector dimension config since it is figured out by langchain integration
- added comment to make Composite index instead
- removed some false advantages mentioned in markdowns of Bhive

VirajAgarwal1 committed:
update: gsi tutorial

- Improved section headings for better navigation and understanding.
- Added performance comparision of crew ai agent with and without gsi index.
- Added detailed explanations for setup, prerequisites, and GSI vector search..
- Streamlined content for readability and coherence.

VirajAgarwal1 committed:
update: gsi tutorial

- performance testing to be simpler and without using crewai, just pure rag
- mentioned composite index much more explicitily
